src/test/java/io/airlift/slice/SliceUtf8Benchmark.java

```diff
@@ -0,0 +1,282 @@
+@State(Thread)
+@OutputTimeUnit(NANOSECONDS)
+@BenchmarkMode(AverageTime)
+@Fork(1)
+@Warmup(iterations = 4, time = 500, timeUnit = MILLISECONDS)
+@Measurement(iterations = 5, time = 500, timeUnit = MILLISECONDS)
+public class SliceUtf8Benchmark
+{
+    @Benchmark
+    public int benchmarkLengthOfCodePointFromStartByte(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int i = 0;
+        int codePoints = 0;
+        while (i < slice.length()) {
+            i += lengthOfCodePointFromStartByte(slice.getByte(i));
+            codePoints++;
+        }
+        if (codePoints != data.getLength()) {
+            throw new AssertionError();
+        }
+        return codePoints;
+    }
+
+    @Benchmark
+    public int benchmarkCountCodePoints(BenchmarkData data)
+    {
+        int codePoints = countCodePoints(data.getSlice());
+        if (codePoints != data.getLength()) {
+            throw new AssertionError();
+        }
+        return codePoints;
+    }
+
+    @Benchmark
+    public int benchmarkOffsetByCodePoints(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int offset = offsetOfCodePoint(slice, data.getLength() - 1);
+        if (offset + lengthOfCodePoint(slice, offset) != slice.length()) {
+            throw new AssertionError();
+        }
+        return offset;
+    }
+
+    @Benchmark
+    public Slice benchmarkSubstring(BenchmarkData data)
+    {
+        Slice slice = data.getSlice();
+        int length = data.getLength();
+        return substring(slice, (length / 2) - 1, length / 2);
+    }
+
+    @Benchmark
+    public Slice benchmarkReverse(BenchmarkData data)
+    {
+        return reverse(data.getSlice());
+    }
+
+    @Benchmark
+    public Slice benchmarkToLowerCase(BenchmarkData data)
+    {
+        return toLowerCase(data.getSlice());
+    }
+
+    @Benchmark
+    public Slice benchmarkToUpperCase(BenchmarkData data)
+    {
+        return toUpperCase(data.getSlice());
+    }
+
+    @Benchmark
+    public Slice benchmarkLeftTrim(WhitespaceData data)
+    {
+        return leftTrim(data.getLeftWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkRightTrim(WhitespaceData data)
+    {
+        return rightTrim(data.getRightWhitespace());
+    }
+
+    @Benchmark
+    public Slice benchmarkTrim(WhitespaceData data)
+    {
+        return trim(data.getBothWhitespace());
+    }
+
+
+    @State(Thread)
+    public static class BenchmarkData
+    {
+        private static final int[] ASCII_CODE_POINTS;
+        private static final int[] ALL_CODE_POINTS;
+
+        static {
+            ASCII_CODE_POINTS = new int[0x7F];
+            for (int i = 0; i < ASCII_CODE_POINTS.length; i++) {
+                ASCII_CODE_POINTS[i] = i;
+            }
+
+            int[] codePoints = new int[MAX_CODE_POINT];
+            int index = 0;
+            for (int codePoint = 0; codePoint < MAX_CODE_POINT; codePoint++) {
+                if (Character.isDefined(codePoint)) {
```

Comment:
Looks like this code is not up to date. As is return, surrogate characters will be included.